_algorithm_design/on_computing_next_greater_intervals_in_linear_time.md

```diff
@@ -0,0 +1,179 @@
+    - Thus instead of searching linearly leftwards, we could skip the elements in between and directly check $list[leftBound[N-1]]$, then $list[leftBound[leftBound[N-1]]]$ and so on until we find the element greater than $curr$.
+    - The set of indices $list[N-1]$, $list[leftBound[N-1]]$, $list[leftBound[leftBound[N-1]]]$ and so on form a topology of ancestors dictated strictly by ordering with respect to the $>$ operator. At this point we can safely conclude that computation of left bound for a new element depends strictly on the topological ancestors of the last element in the $list$.
+
+Hence we have pruned the search space but by how much? We will discuss the answer to this shortly.
+
+#### The monotonic stack
+
+There are two noteworthy observations to make considering that we need only maintain the topological ancestors of the last element in the $list$ to compute the left bound of $curr$:
+- case1 tells us that $curr$ can simply be added to the front of the topology without disturbing the rest of its structure. case2 tells us that once we find the left bound of $curr$, we don't need to maintain any elements between $curr$ and its left bound. If you stare at the screen long enough, you'll notice that case1 and case2 together manifest a `last-in-first-out` behaviour in the topology of ancestors. Hence a stack can be used to represent the topological ancestors of the last element in $list$.
+- case2 also suggests that the topology obeys the following invariant $ancestor(item) > item \, \forall item \in topology$. The invariant is equivalent to monotonicity in the stack.
+
+Thus monotonic stack gets its name from the two observations made above. So far so good! Now we are in a position to write down the semantics of the proposed data structure. I am (again) choosing Pythonic pseudo code to write the semantics.
+
+{% highlight python %}
+class MonotonicStack(Stack):
+
+    @abstractmethod
+    def is_empty():
+        """Returns true if the topology doesn't contain any elements, false otherwise"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def push(idx):
+        """Adds idx to top of the topology"""
+        raise NotImplementedError
+    
+    @abstractmethod
+    def pop():
+        """Removes an element from the top of the topology"""
+        raise NotImplementedError
+    
+    @abstractmethod
+    def top():
+        """Returns the element at the top of the topology"""
+        raise NotImplementedError
+{% endhighlight %}
+
+Let's try to write an algorithm to compute the left bound for every element in the list using the observations and specified semantics.
+
+{% highlight python %}
+topo = MonotonicStack()
+for i, item in enumerate(list):
+    if topo.is_empty() or list[topo.get_next()] > item:
```

Comment:
topo.get_next(), what is this method ? no explaination during the semantics, i assume you meant top() here

Reply:
Yeah yeah. It should be `top()`. Thanks for pointing out.